src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/ModelExplorer.cs

```diff
@@ -0,0 +1,378 @@
+        /// <remarks>
+        /// <para>
+        /// A <see cref="ModelExplorer"/> created by 
+        /// <see cref="GetExplorerForExpression(Type, Func{object, object})"/>
+        /// represents the result of executing an arbitrary expression against the model contained
+        /// in the current <see cref="ModelExplorer"/> instance.
+        /// </para>
+        /// <para>
+        /// The returns <see cref="ModelExplorer"/> will have the current instance set as its <see cref="Container"/>.
+        /// </para>
+        /// </remarks>
+        public ModelExplorer GetExplorerForExpression([NotNull] Type modelType, Func<object, object> modelAccessor)
+        {
+            var metadata = _metadataProvider.GetMetadataForType(modelType);
+            return GetExplorerForExpression(metadata, modelAccessor);
+        }
+
+        /// <summary>
+        /// Gets a <see cref="ModelExplorer"/> for the provided model value and model <see cref="Type"/>.
+        /// </summary>
+        /// <param name="modelType">The model <see cref="Type"/>.</param>
+        /// <param name="model">The model value.</param>
+        /// <returns>A <see cref="ModelExplorer"/>.</returns>
+        /// <remarks>
+        /// <para>
+        /// A <see cref="ModelExplorer"/> created by 
+        /// <see cref="GetExplorerForExpression(ModelMetadata, Func{object, object})"/>
+        /// represents the result of executing an arbitrary expression against the model contained
+        /// in the current <see cref="ModelExplorer"/> instance.
+        /// </para>
+        /// <para>
+        /// The returns <see cref="ModelExplorer"/> will have the current instance set as its <see cref="Container"/>.
+        /// </para>
+        /// </remarks>
+        public ModelExplorer GetExplorerForExpression([NotNull] ModelMetadata metadata, Func<object, object> modelAccessor)
+        {
+            return new ModelExplorer(_metadataProvider, this, metadata, modelAccessor);
+        }
+
+        private ModelMetadata GetMetadataForRuntimeType()
+        {
+            // We want to make sure we're looking at the runtime properties of the model, and for
+            // that we need the model metadata of the runtime type.
+            var metadata = Metadata;
+            if (Metadata.ModelType != ModelType)
+            {
+                metadata = _metadataProvider.GetMetadataForType(ModelType);
+            }
+
+            return metadata;
+        }
+
+        private ModelExplorer CreateExplorerForProperty(
+            ModelMetadata propertyMetadata, 
+            PropertyHelper propertyHelper)
+        {
+            if (propertyHelper == null)
+            {
+                return new ModelExplorer(_metadataProvider, this, propertyMetadata, modelAccessor: null);
+            }
+
+            var modelAccessor = new Func<object, object>((c) =>
+            {
+                return c == null ? null : propertyHelper.GetValue(c);
```

Comment:
general nit: preference for one- and two-letter lambda parameters makes changes harder to read